client/src/main/scala/org/http4s/client/middleware/Retry.scala

```diff
@@ -2,71 +2,141 @@ package org.http4s
-      prepareLoop(req.withEmptyBody, attempts + 1).schedule(duration)
+    def nextAttempt(req: Request, attempts: Int, duration: FiniteDuration, retryHeader: Option[`Retry-After`]): Task[DisposableResponse] = {
+      val headerDuration = retryHeader.map { h =>
+        h.retry match {
+          case Left(d) => Instant.now().until(d.toInstant, ChronoUnit.SECONDS)
+          case Right(secs) => secs
+        }
+      }.getOrElse(0L)
+      val sleepDuration = Math.max(headerDuration, duration.length).seconds
+      prepareLoop(req.withEmptyBody, attempts + 1).schedule(sleepDuration)
     }
 
     client.copy(open = Service.lift(prepareLoop(_, 1)))
   }
 }
 
-
 object RetryPolicy {
+  /** Decomposes a retry policy into components that are typically configured
+    * individually.
+    * 
+    * @param backoff a function of attempts to an optional
+    * FiniteDuration.  Return None to stop retrying, or some
+    * duration after which the request will be retried.  See
+    * `exponentialBackoff` for a useful implementation.
+    * 
+    * @param retriable determines whether the request is retriable
+    * from the request and either the throwable or response that was
+    * returned.  Defaults to `defaultRetriable`.
+    */
+  def apply(
+    backoff: Int => Option[FiniteDuration],
+    retriable: (Request, Either[Throwable, Response]) => Boolean = defaultRetriable
+  ): RetryPolicy = { (req, result, retries) =>
+    if (retriable(req, result)) backoff(retries)
+    else None
+  }
+
+  /** Statuses that are retriable, per HTTP spec */
+  val RetriableStatuses = Set(
+    RequestTimeout,
+    // TODO Leaving PayloadTooLarge out until we model Retry-After
+    InternalServerError,
+    ServiceUnavailable,
+    BadGateway,
+    GatewayTimeout
+  )
+
+  /** Default logic for whether a request is retriable.  Returns true if
+    * the request method does not permit a body and the result is
+    * either a throwable or has one of the `RetriableStatuses`.
+    * 
+    * Caution: more restrictive than 0.16.  That would inspect the
+    * body for effects, and resumbmit only if the body was pure (i.e.,
+    * only emits and halts).  The fs2 algebra does not let us inspect
+    * the stream for effects, so we can't safely resubmit.  For the
+    * old behavior, use [[unsafeRetriable]].  To ignore the response
+    * codes, see [[recklesslyRetriable]].
+    */
+  def defaultRetriable(req: Request, result: Either[Throwable, Response]): Boolean = {
+    if (req.method.isInstanceOf[Method.NoBody]) isErrorOrRetriableStatus(result)
+    else false
+  }
+
+  /** Returns true if the request method is idempotent and the result is
+    * either a throwable or has one of the `RetriableStatuses`.  This is
+    * the `defaultRetriable` behavior from 0.16.
+    * 
+    * Caution: if the request body is effectful, the effects will be
+    * run twice.  The most common symptom of this will be resubmitting
+    * an empty request body.
```

Comment:
This is a really horrid thing to do, but some people will want this.  An alternative is to buffer request bodies, which would slow things down.